File: src/main/java/dev/kensa/state/TestState.java
package dev.kensa.state;

public enum TestState {
    Disabled("Disabled"),
    Failed("Failed"),
    NotExecuted("Not Executed"),
    Passed("Passed");

    private final String description;

    TestState(String description) {
        this.description = description;
    }

    public String description() {
        return description;
    }

    public TestState overallStateFrom(TestState otherState) {
        if(this == Failed || this == Disabled || otherState == NotExecuted || otherState == Disabled) {
            return this;
        }

        if(otherState == Failed || otherState == Passed) {
            return otherState;
        }

        return this;
    }
}

File: core/src/main/kotlin/dev/kensa/state/TestState.kt
package dev.kensa.state

import java.util.*

enum class TestState(val description: String) {
    Disabled("Disabled"),
    Failed("Failed"),
    NotExecuted("Not Executed"),
    Passed("Passed");

    fun overallStateFrom(otherState: TestState): TestState =
            if (DISABLED_FAILED.contains(this) || DISABLED_NOT_EXECUTED.contains(otherState)) {
                this
            } else if (FAILED_PASSED.contains(otherState)) {
                otherState
            } else {
                this
            }

    companion object {
        val DISABLED_FAILED: EnumSet<TestState> = EnumSet.of(Disabled, Failed)
        val DISABLED_NOT_EXECUTED: EnumSet<TestState> = EnumSet.of(Disabled, NotExecuted)
        val FAILED_PASSED: EnumSet<TestState> = EnumSet.of(Failed, Passed)
    }
}